drivers/sds011/sds011.c

```diff
@@ -0,0 +1,351 @@
+            res = SDS011_OK;
+        }
+        else {
+            res = SDS011_INVALID_RESPONSE;
+        }
+    }
+    else {
+        res = SDS011_INVALID_CHKSUM;
+    }
+
+    /* reset mutex state */
+    mutex_unlock(&dev->cb_lock);
+
+    /* if no active reporting callback is registered, UART can be disabled */
+    if((dev->cb == NULL) &&
+       (uart_init(dev->params.uart, SDS011_UART_BAUDRATE, NULL, NULL) != 0)) {
+        res = SDS011_ERROR;
+    }
+
+    /* release device */
+    mutex_unlock(&dev->dev_lock);
+
+    return res;
+}
+
+/** @brief shorthand for getting a single byte property with _send_recv_cmd  */
+static int _get_property(sds011_t *dev, uint8_t *cmd, size_t len,
+                         uint8_t *p, uint8_t p_idx)
+{
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, len, recv);
+
+    if (res == SDS011_OK) {
+        *p = recv[p_idx];
+    }
+
+    return res;
+}
+
+int sds011_init(sds011_t *dev, const sds011_params_t *params)
+{
+    assert(params->uart < UART_NUMOF);
+
+    if ((params->pwr_pin != GPIO_UNDEF) &&
+        (gpio_init(params->pwr_pin, GPIO_OUT) != 0)) {
+        return SDS011_ERROR;
+    }
+
+    memcpy(&dev->params, params, sizeof(sds011_params_t));
+
+    mutex_init(&dev->dev_lock);
+    mutex_init(&dev->cb_lock);
+
+    dev->cb = NULL;
+
+    sds011_power_on(dev);
+
+    return SDS011_OK;
+}
+
+int sds011_register_callback(sds011_t *dev, sds011_callback_t cb, void *ctx)
+{
+    mutex_lock(&dev->dev_lock);
+    dev->cbctx = ctx;
+    dev->cb = cb;
+
+    /* either register un unregister the uart callback */
+    if (uart_init(dev->params.uart, SDS011_UART_BAUDRATE,
+                  cb == NULL ? NULL : _rx_cb,
+                  cb == NULL ? NULL : dev) != 0) {
+        mutex_unlock(&dev->dev_lock);
+        return SDS011_ERROR;
+    }
+    mutex_unlock(&dev->dev_lock);
+    return SDS011_OK;
+}
+
+void sds011_power_on(const sds011_t *dev)
+{
+    if(dev->params.pwr_pin != GPIO_UNDEF) {
+        gpio_write(dev->params.pwr_pin, dev->params.pwr_ah);
+    }
+}
+
+void sds011_power_off(const sds011_t *dev)
+{
+    if(dev->params.pwr_pin != GPIO_UNDEF) {
+        gpio_write(dev->params.pwr_pin, !dev->params.pwr_ah);
+    }
+}
+
+int sds011_get_reporting_mode(sds011_t *dev, sds011_reporting_mode_t *mode)
+{
+    uint8_t cmd[] = {SDS011_CMD_DB1_SET_DR_MODE, SDS011_CMD_OPT_QUERY};
+    return _get_property(dev, cmd, sizeof(cmd), mode, SDS011_DB3_IDX);
+}
+
+int sds011_set_reporting_mode(sds011_t *dev, sds011_reporting_mode_t mode)
+{
+    uint8_t cmd[] = {SDS011_CMD_DB1_SET_DR_MODE, SDS011_CMD_OPT_SET, mode};
+
+    uint8_t recv[SDS011_FRAME_RECV_LEN];
+    int res = _send_recv_cmd(dev, cmd, sizeof(cmd), recv);
+
+    if (res == SDS011_OK) {
+        if ((recv[SDS011_DB2_IDX] == SDS011_CMD_OPT_SET) &&
+            (recv[SDS011_DB3_IDX] == mode)) {
+            return SDS011_OK;
+        }
+        else{
+            return SDS011_ERROR;
```

Comment:
see above, I suggest to remove this `else` and simply return `SDS011_ERROR` by default when not successful

Reply:
done